test/QApiTests.h

```diff
@@ -0,0 +1,220 @@
+#pragma once
+#include <gtest/gtest.h>
+#include <string>
+#include <chrono>
+#include <vector>
+#include <atomic>
+#include <iostream>
+#include <future>
+
+namespace QApiTests {
+   using ResultType = std::vector<std::string>;
+
+
+   template <typename Sender>
+   ResultType Push(Sender q, size_t start, size_t stop,
+                   std::atomic<bool>& producerStart, std::atomic<bool>& consumerStart) {
+      using namespace std::chrono_literals;
+
+      std::vector<std::string> expected;
+      expected.reserve(stop - start);
+      producerStart.store(true);
+      while (!consumerStart.load()) {
+         std::this_thread::sleep_for(1us);
+      }
+
+      for (auto i = start; i < stop; ++i) {
+         std::string value = std::to_string(i);
+         expected.push_back(value);
+         while (false == q.push(value)) {
+            std::this_thread::sleep_for(1us); // // yield is too aggressive
+         }
+      }
+      std::ostringstream oss;
+      oss << "Push Until: " << q.mStats.FlushAsString() << std::endl;
```

Comment:
updated.

Reply:
the `mStats` usage is to show the API access. It's a completely transparent wrapper struct with all the member variables exposed on purpose.  I.e. there are more function calls you can do through the `mStats` object. Potentially the API of `mStats` will also grow in the future ... in which case it would be tedious and quite terrible to keep updating this API.

Reply:
Personally I don't like the naming of `m` prefix for structs such as these. It would look cleaner if it would be 
`q.stats_.FlushAsString` but that's not following our coding guidelines.